platform/engineering/request-for-change/2020-01-08-ssoe-removing-implicit-mvi-queries.md

```diff
@@ -0,0 +1,106 @@
+
+2. **526 Add Person Requests**
+The 526 (Disability compensation) application is currently adding a feature that performs an MVI `Add Person` request to establish several new identifiers for a user in MVI. Performing an `Add Person` requires a search token that can be obtained by performing an MVI query for a user. The token has a short expiry, meaning the `Add Person` request must be issued within 60 seconds of the MVI query.<br/>
+So, regardless of the SSOe integration, this feature would require the ability to perform an explicit MVI lookup and ignore any cached results (because of the search token expiry). So the proposal is again to perform an explicit MVI lookup. If the 526 application needs to consume the new identifiers, then it likewise can update the `user_identity` cache (aka the session) with the relevant identifiers obtained from the lookup + add person process.
+
+3. **Healthcare Application MVI attr_service**
+The healthcare application implements a totally parallel MVI service with a separate configuration from the one used by vets-api at large. This is really a development environment workaround, to solve the problem of unreconcilable environment mappings between vets-api, MVI, and the HCA system. This application performs an explicit MVI lookup using its own service configuration and will not be impacted by any of the above changes.
+
+4. **(Future) Delegation Lookups**
+VA.gov is likely to introduce delegation features powered by MVI in the future. These let user A delegate access to their information to user B. Record of that delegation is recorded in MVI, and a flag on a user record indicates whether someone is a delegate.<br/>
+The isDelegate flag is present in SSOe SAML assertions, so the login flow an account for delegation. The next step of selecting a delegatee on behalf of whom to act, will likely require one or more additional explicit MVI requests which likewise should be separated  from the Redis cache_aside mechanism.
+
+5. **OAuth Flow - Lighthouse**
+SAML authentication and MVI resolution is also performed in the context of the Lighthouse OAuth flow. MVI is queried by an API endpoint triggered by the SAML proxy, and cached MVI data is used when serving OAuth-protected API requests.<br/>
+At this point, there is no intent to integrate SSOe with the OAuth SAML interaction. Signing in to authorize an OAuth application will not establish an IAM session.<br/>
+Current proposal for the OAuth code path is to leave the MVI caching mechanism in place. This will mean that the `User` model definitely needs to remain robust to `va_profile` being populated either from an SSOe SAML payload or from an MVI lookup for the indefinite future. We should audit all of the OAuth-protected APIs to reconcile uses of various delegated methods on `User` to make sure there are no conflicts.
```

Comment:
You've noted below that vets-api prefers ID.me values for some fields. Ignoring those, thinking just about the fields that have a "correct" value like `icn` and `dslogon_edipi`, is there any reason to expect `va_profile` to contain different values?

Reply:
ICN might come along with an MHV credential - but in that case we look up from MVI using that ICN as the primary key.  Likewise EDIPI might come along with a DSLogon credential, but in that case we use the EDIPI as the lookup key. So I don't see any case where those identifiers would not match what comes back in va_profile.